laelf: Add reloc types for LA32 (alternative version allowing separating pcaddu12i and addi.w/ld.[bhw])

xry111 opened the pull request:
This is an alternative of #14, to remove the restriction that the `addi.w` or `ld.[bhw]` instruction must immediately follow the corresponding `pcaddu12i` instruction.

@flygoat believes we should remove the restriction for flexibly.  I'm unsure (feeling a dilemma) so I'm posting both alternatives for review.

Cc @heiher @chenhuacai @xen0n

---

LA32 (both R and S) implementations lack pcaddu18i, so we cannot use R_LARCH_CALL36.  Add `R_LARCH_CALL32` for medium code model function call on LA32.

And, LA32R implementations may lack pcalau12i, so we cannot use the existing reloc types assuming pcalau12i for LA32R.  Add 4 reloc types for the pcaddu12i instruction:

- `R_LARCH_PCADD_HI20`, for la.pcrel
- `R_LARCH_PCADD_GOT_HI20`, for la.got
- `R_LARCH_PCADD_TLS_IE_HI20`, for la.tls.ie
- `R_LARCH_PCADD_TLS_DESC_HI20`, for la.tls.desc

and one reloc type, `R_LARCH_PCADD_LO12_I`, for the addi.w or ld.[bhw] instruction, to be paired with those 4 reloc types.

The existing R_LARCH_TLS_LE_HI20_R and R_LARCH_TLS_LE_LO12_R are enough for la.tls.le.  We won't support la.tls.{ld,gd} on LA32R as there's no reason to use them instead of la.tls.desc.

FlyGoat commented:
Thanks! I'd prefer this version.

FYI there are a couple of OoO LA32R cores, the most famous one seems to be: https://github.com/LainChip/LainCore, so this flexibility means a lot!

jiegec commented:
I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.

chenhuacai commented:
> I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.

Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.

FlyGoat commented:
I’ll be working on LLVM side.

cloudspurs commented:
> > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> 
> Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.

How is the progress of the binutls patch? We are also preparing to implement a binutils patch based on PR16.

jiegec commented:
> > > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> > 
> > 
> > Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.
> 
> How is the progress of the binutls patch? We are also preparing to implement a binutils patch based on PR16.

Not much progress, I have only sorted out the DWARF alignment problem, but the relocations are not added. I am working on GCC most recently.

If you are interested, please go ahead. And if you have any progress ahead of mime, I am happy to test it.

cloudspurs commented:
> > > > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> > > 
> > > 
> > > Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.
> > 
> > 
> > How is the progress of the binutls patch? We are also preparing to implement a binutils patch based on PR16.
> 
> Not much progress, I have only sorted out the DWARF alignment problem, but the relocations are not added. I am working on GCC most recently.
> 
> If you are interested, please go ahead. And if you have any progress ahead of mime, I am happy to test it.

I will add these relocations as soon as possible.

Luffy-tsai commented:
> > > > > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> > > > 
> > > > 
> > > > Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.
> > > 
> > > 
> > > How is the progress of the binutls patch? We are also preparing to implement a binutils patch based on PR16.
> > 
> > 
> > Not much progress, I have only sorted out the DWARF alignment problem, but the relocations are not added. I am working on GCC most recently.
> > If you are interested, please go ahead. And if you have any progress ahead of mime, I am happy to test it.
> 
> I will add these relocations as soon as possible.

I've added these relocations and the implementation is based on RISCV.
https://github.com/loongson-cailulu/binutils-gdb/tree/la32r

xry111 commented:
> > > > > > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> > > > > 
> > > > > 
> > > > > Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.
> > > > 
> > > > 
> > > > How is the progress of the binutls patch? We are also preparing to implement a binutils patch based on PR16.
> > > 
> > > 
> > > Not much progress, I have only sorted out the DWARF alignment problem, but the relocations are not added. I am working on GCC most recently.
> > > If you are interested, please go ahead. And if you have any progress ahead of mime, I am happy to test it.
> > 
> > 
> > I will add these relocations as soon as possible.
> 
> I've added these relocations and the implementation is based on RISCV. https://github.com/loongson-cailulu/binutils-gdb/tree/la32r

Thanks very much!  As this provides more functionality than #14 and we have a working implementation, should we just close #14 and focus on this now?

Luffy-tsai commented:
> > > > > > > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> > > > > > 
> > > > > > 
> > > > > > Good. I prefer PR16 than PR14, and I hope your code contribution is helpful to confirm the completeness of this PR.
> > > > > 
> > > > > 
> > > > > How is the progress of the binutls patch? We are also preparing to implement a binutils patch based on PR16.
> > > > 
> > > > 
> > > > Not much progress, I have only sorted out the DWARF alignment problem, but the relocations are not added. I am working on GCC most recently.
> > > > If you are interested, please go ahead. And if you have any progress ahead of mime, I am happy to test it.
> > > 
> > > 
> > > I will add these relocations as soon as possible.
> > 
> > 
> > I've added these relocations and the implementation is based on RISCV. https://github.com/loongson-cailulu/binutils-gdb/tree/la32r
> 
> Thanks very much! As this provides more functionality than #14 and we have a working implementation, should we just close #14 and focus on this now?

I think we can focus on this, and we currently plan to continue to implement it.

cloudspurs commented:
> I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.

Can the gcc loongarch32-new branch complete the compilation? Some ICEs occurred in my environment.

jiegec commented:
> > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> 
> Can the gcc loongarch32-new branch complete the compilation? Some ICEs occurred in my environment.

Not yet, I am trying to fix it, but I have not figured out how.

cloudspurs commented:
> > > I have started rebasing my previous loongarch32 upstreaming attempt based on this pr (https://github.com/jiegec/binutils-gdb/tree/loongarch32, https://github.com/jiegec/gcc/tree/loongarch32 and https://github.com/jiegec/glibc/tree/loongarch32) at https://github.com/jiegec/gcc/tree/loongarch32-new and https://github.com/jiegec/binutils-gdb/tree/loongarch32-new.
> > 
> > 
> > Can the gcc loongarch32-new branch complete the compilation? Some ICEs occurred in my environment.
> 
> Not yet, I am trying to fix it, but I have not figured out how.

This (https://github.com/cloudspurs/gcc/tree/la32) can complete the gcc-first compilation with build-many-glibcs.sh. Mostly based on your loongarch32-new branch.

jiegec commented:
> 
> This (https://github.com/cloudspurs/gcc/tree/la32) can complete the gcc-first compilation with build-many-glibcs.sh. Mostly based on your loongarch32-new branch.

Thanks a lot! I am quite busy these days. If you are interested, you can take over the upstreaming efforts.

cloudspurs commented:
https://github.com/cloudspurs/binutils-gdb/tree/la32，
https://github.com/cloudspurs/gcc/tree/la32，
https://github.com/cloudspurs/glibc/tree/la32。
These can build a complete loongarch32 toolchain with build-many-glibcs.py.
But no testing has been done yet.

cloudspurs commented:
Use qemu to run a LA32 linux system:
https://github.com/sunhaiyong1978/CLFS-for-LoongArch32/blob/main/Qemu_For_LoongArch32-Simple.md

heiher commented on laelf.adoc:
```diff
@@ -815,6 +815,38 @@ See <<code_models>> for how it works on various code models.
 |`+(*(uint32_t *) PC) [24 ... 5] = (GOT+GD) [21 ... 2]+`
+
+|127
+|R_LARCH_CALL32
```
The relocation type `R_LARCH_CALL36` uses 36 effective bits for operand encoding, corresponding to a signed 38-bit PC-relative offset. From the perspective of semantic consistency, wouldn't it make more sense to name `R_LARCH_CALL32` as `R_LARCH_CALL30` instead?

xry111 replied:
I guess so.  @cloudspurs: if you agree I'll change it.

cloudspurs replied:
I also think so, please change it, thanks.

heiher commented:
LLVM WIP: https://github.com/heiher/llvm-project/commits/la32

SixWeining commented on laelf.adoc:
```diff
@@ -815,6 +815,38 @@ See <<code_models>> for how it works on various code models.
+|`+(*(uint32_t *) PC) [24 ... 5] = (GOT+DESC-PC+0x800) [31 ... 12],+`
+
+|132
+|R_LARCH_PCADD_LO12_I
```
Is the `_I` suffix unnecessary?

xen0n replied:
It's for "indirect", not anything like instruction format specifier (e.g. RISC-V "I" which I bet a lot of people will get misled by) or "GPR". This is easily misleading so I'd suggest something more descriptive e.g. "_IND" instead.

xen0n replied:
In our case, most other relocs don't use the indirection semantics introduced here, so it can be very helpful to have an indication like this. We're unlike e.g. RISC-V where all low-part reloc types (IIRC) are of this type.

SixWeining reviewed (commented):
Thanks. Basically it looks good to me. Could you annotate assembly modifiers for these relocations like what #15 do?

heiher commented on laelf.adoc:
```diff
@@ -815,6 +815,38 @@ See <<code_models>> for how it works on various code models.
+|132
+|R_LARCH_PCADD_LO12_I
+|[11 ... 0] bits of 32-bit PC-relative offset to the target of an `R_LARCH_PCADD_*_HI20` reloc filling the immediate slot for a `pcaddu12i` instruction at `S`; the addend of `R_LARCH_PCADD_LO12_I` must be 0
+|`+(*(uint32_t *) PC) [21 ... 10] = (R(S)-S) [11 ... 0],+`
```
> the addend of `R_LARCH_PCADD_LO12_I` must be 0

Why is it required that the `addend` must be `0`?

It's quite common for LLVM to optimize relocations in object files by replacing references to local symbols with section-relative expressions (section + offset).

```
    .text
main:
    nop
.Lpcadd_hi:
    pcaddu12i $a0, %pcadd_hi20(sym)
    addi.w $a0, $a0, %pcadd_lo12(.Lpcaddhi)
```

Local symbol:

```
addi.w	$a0, $a0, 0
0000000c:  R_LARCH_PCADD_LO12_I	.Lpcadd_hi
```

Section+offset:

```
addi.w	$a0, $a0, 0
0000000c:  R_LARCH_PCADD_LO12_I	.text+0x4
```

SixWeining replied:
I was confused about the same thing yesterday, but now I get it. The real addend should already be included in `R(S)`. Right?

heiher replied:
Just to clarify. we need distinguish between the roles of `R_LARCH_PCADD_HI20` and `R_LARCH_PCADD_LO12_I` relocations and how their addends are used.

- `R_LARCH_PCADD_HI20` targets the symbol `sym + off`, and the `addend` represents that `off`. It's part of the final address calculation for `sym`.
- On the other hand, `R_LARCH_PCADD_LO12_I` is used to reference the address of the corresponding `pcaddu12i` instruction (.Lpcadd_hi). its purpose is to locate that instruction precisely.

When LLVM replaces the `.Lpcadd_hi` label with a section-relative expression like `.text + 4`, it effectively eliminates the need to keep `.Lpcadd_hi` as a symbol in the object file, reducing symbol table size.

So what we typically see in the objet file is:

```
R_LARCH_PCADD_LO12_I:  .text + 4 (addend = 4)
```

heiher replied:
Unless there is a specific reason to enforce it, I would recommend relaxing this. It could either be removed or rephrased as:

the addend of `R_LARCH_PCADD_LO12_I` is used solely to locate the paired `pcaddu12i` instruction and does not participate in calculating the final target address of the `R_LARCH_PCADD_*_HI20` relocation.

cloudspurs replied:
section + offset format is difficultly to do linker relaxation.

heiher replied:
Is the reason for requiring `addend == 0` in the ABI specs mainly to support linker relaxation?

If so, wouldn't it make sense to allow section+offset in general, and only enforce symbol-based references when relaxation is enabled?

cloudspurs replied:
relaxed .o and norelaxed .o may cause ld error ".eh_frame_hdr refers to overlapping FDEs".
https://github.com/bminor/binutils-gdb/blob/master/gas/config/tc-loongarch.h#L71

heiher replied:
So, non-relaxed .o files are unlikely to be used in practice? If that's the case, it make sense that relaxing the restriction wouldn't be very useful. I'm okay with leaving it as is.

cloudspurs committed:
Add LA32S and LA32R relocations

127 R_LARCH_CALL30
128 R_LARCH_PCADD_HI20
129 R_LARCH_PCADD_LO12
130 R_LARCH_GOT_PCADD_HI20
131 R_LARCH_GOT_PCADD_LO12
132 R_LARCH_TLS_IE_PCADD_HI20
133 R_LARCH_TLS_IE_PCADD_LO12
134 R_LARCH_TLS_LD_PCADD_HI20
135 R_LARCH_TLS_LD_PCADD_LO12
136 R_LARCH_TLS_GD_PCADD_HI20
137 R_LARCH_TLS_GD_PCADD_LO12
138 R_LARCH_TLS_DESC_PCADD_HI20
139 R_LARCH_TLS_DESC_PCADD_LO12

xry111 commented:
To simplify the linker implementation, we introduced one HI reloc for each LO reloc.  And to allow assembling the existing assembly code using la.tls.{gd/ld}, we had to add back LD/GD relocs :(.

SixWeining commented:
Done in the latest release (v2.50). Thank you all.
